### PR TITLE
Infer the essential sources tar.gz file name instead of mentioning it…

### PR DESCRIPTION
**Pull request by alanoe**
… explicitly in package yaml file

When creating the essential sources tar.gz from a git repository, there is no reason to use a
file name different from the package name, so this attribute is useless in this case.

When downloading the essential sources tar.gz instead of creating it from a git repository,
expects_source attribute may be useful. This currently happens in gcc, kubernetes and libguestfs.
In kubernetes and libguestfs, a tar.gz is downloaded using wget, so expects_source was
dropped by passing an output file name in standard format used by OpenPower Host OS build
script to the wget command. In gcc, a tar.bz2 is created instead of a tar.gz file, so the
expects_source attribute must be kept. It should be evaluated if using tar.bz2 instead of
tar.gz in gcc is essential and, if it is not, drop it for good in a future patch.

### CODE REVIEW

**Comment by olavphibm** [comment]
This depends on open-power-host-os/builds#132

**Comment by olavphibm** [comment]
start tests

**Comment by fabianorosas** [comment]
start pylint

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

**Comment by alanoe** [comment]
start build

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

File: libseccomp/libseccomp.yaml [hunk 1]
```diff
@@ -3,7 +3,6 @@ Package:
  clone_url: 'https://github.com/seccomp/libseccomp.git'
  branch: 'release-2.3'
  commit_id: 'eedd26dc59641878dabd771e2ff15e729a85ac69'
- expects_source: "libseccomp-2.3.1"
```
Comment (olavphibm) [inline libseccomp/libseccomp.yaml]: The `%{version}` tag needs to be removed from the spec file.

Comment (olavphibm) [inline libseccomp/libseccomp.yaml]: The `%{version}` tag needs to be removed from the spec file.

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

**Comment by olavphibm** [comment]
start build

**Comment by mopsfelder** [comment]
This is a nice enhancement.

I noticed that spec files are not uniform, some define `Source` as `%{name}.tar.gz` and others don't, using the literal package name.

What if we unify `Source` as just `%{name}.tar.gz` in all spec files?

**Comment by alanoe** [comment]
OH> Please make sure the Source0 and %setup fields are matching the expects_source default

Done.

MO> What if we unify Source as just %{name}.tar.gz in all spec files?

I agree, but I prefer to do it in a separate patch.

File: gingerbase/gingerbase.yaml [hunk 2]
```diff
@@ -5,7 +5,6 @@ Package:
       src: 'https://github.com/open-power-host-os/gingerbase.git'
       branch: 'hostos-devel'
       commit_id: '82dd2bea6900e3bda3cd14ef6835586b773fc6c8'
-      archive: 'ginger-base'
```
Comment (olavphibm) [inline gingerbase/gingerbase.yaml]: This is still needed, since the name of the package is "ginger-base", but the name of the repository is "gingerbase"

File: kernel/centOS/7.2/kernel.spec [hunk 3]
```diff
@@ -704,7 +704,7 @@ exit 1
 
 tar xzf %{SOURCE0}
 mv linux-source kernel-%{KVRA}
```
Comment (olavphibm) [inline kernel/centOS/7.2/kernel.spec]: I believe this should be: `mv %{name} kernel-%{KVRA}`.
The yaml file also needs to be changed.

File: linux-firmware/centOS/7.2/linux-firmware.spec [hunk 4]
```diff
@@ -237,7 +237,7 @@ contained inside the provided LICENSE file. Please read it carefully.
 
 %prep
 # Setup Source0 and additional Source1
-%setup -q -c -n linux-firmware-%{checkout}
```
Comment (olavphibm) [inline linux-firmware/centOS/7.2/linux-firmware.spec]: How was this even working?

**Comment by fabianorosas** [comment]
Relates to open-power-host-os/builds#38

File: gingerbase/gingerbase.yaml (hunk 2 above)
Comment (alanoe) [inline gingerbase/gingerbase.yaml]: I wonder why someone names the git repository and the rpm package differently. =/

An alternative is to rename the git repository to ginger-base to keep names in sync.

A second alternative is to use gingerbase in Source0 definition in rpm spec instead of %{name}. 

I'd prefer to use one of these alternatives in this case (and in any other package if such inconsistency arises again), eliminating completely the need of this 'archive' attribute.

What do you think?

File: kernel/centOS/7.2/kernel.spec (hunk 3 above)
Comment (alanoe) [inline kernel/centOS/7.2/kernel.spec]: Done

File: linux-firmware/centOS/7.2/linux-firmware.spec (hunk 4 above)
Comment (alanoe) [inline linux-firmware/centOS/7.2/linux-firmware.spec]: Because of -c parameter which creates the top-level directory. See http://rpm.org/max-rpm-snapshot/s1-rpm-inside-macros.html

**Comment by alanoe** [comment]
Requested changes done. Pls let us merge this ASAP so I do not need to rebase this patch again.

**Comment by alanoe** [comment]
I could use help to figure out why the build failed, it does not seem related to the patch.

**Comment by fabianorosas** [comment]
> I could use help to figure out why the build failed, it does not seem related to the patch.

https://github.com/open-power-host-os/infrastructure/pull/13

I have edited the job manually so that builds from versions stop failing.

**Comment by olavphibm** [comment]
start build

File: kubernetes/kubernetes.yaml [hunk 5]
```diff
@@ -10,7 +10,6 @@ Package:
  sources:
    - url:
       src: 'https://github.com/kubernetes/kubernetes/archive/4a3f9c5b19c7ff804cbc1bf37a15c044ca5d2353/kubernetes-4a3f9c5.tar.gz'
```
Comment (olavphibm) [inline kubernetes/kubernetes.yaml]: The file downloaded here is named "kubernetes-4a3f9c5.tar.gz". Either we set the `archive` attribute to `kubernetes`, or (better) change the url to a git source, specifying the commit ID. The tar.gz archive being downloaded should be identical to the repository contents, since its created automatically by github.

Comment (alanoe) [inline kubernetes/kubernetes.yaml]: As I mentioned before, imho archive attrib is always a hack, we do not need it. I've changed build code to rename the downloaded file to the file name specified in "archive" (and its default value is <package_name>.tar.gz), to make URL download  consistent with Mercurial/Git downloads.

Changing kubernetes package from a URL source to a git source may be done in a separate patch if we want.

File: libseccomp/centOS/7/libseccomp.spec [hunk 6]
```diff
@@ -6,7 +6,8 @@ Release: 1%{?dist}%{?ibm_release}
 ExclusiveArch: %{ix86} x86_64 %{arm} aarch64 mipsel mips64el ppc64 ppc64le s390 s390x
 License: LGPLv2
 Group: System Environment/Libraries
-Source: https://github.com/seccomp/libseccomp/releases/download/v%{version}/%{name}-%{version}.tar.gz
+# original: https://github.com/seccomp/libseccomp/releases/download/v%{version}/%{name}-%{version}.tar.gz
+Source: %{name}-%{version}.tar.gz
```
Comment (mopsfelder) [inline libseccomp/centOS/7/libseccomp.spec]: What is the benefit of this change?

Fedora advises to keep URL in `Source` as a manner of tracking its origin, see https://fedoraproject.org/wiki/Packaging:SourceURL#Referencing_Source

Comment (alanoe) [inline libseccomp/centOS/7/libseccomp.spec]: IMHO it is *very* anti-intuitive that a URL is accepted in the RPM spec's Source field just for documentation, as nothing is downloaded by rpmbuild, it only uses the URL's file name and looks locally for the source file. Olav agrees with me on this one. I myself got confused with this when editing these rpm spec files and lost time until I found how this worked. Therefore, I disagree with Fedora recommendation. It not because Fedora recommends it that it is good.

If the team wants to keep using this standard, although it is not intuitive, I can undo this change.

File: kubernetes/centOS/7/kubernetes.spec [hunk 7]
```diff
@@ -66,8 +66,8 @@ Summary:        Container cluster management
 License:        ASL 2.0
 URL:            %{import_path}
 ExclusiveArch:  x86_64 ppc64le
-Source0:        https://%{provider_prefix}/archive/%{commit}/%{repo}-%{shortcommit}.tar.gz
-Source1:        https://%{k8s_provider_prefix}/archive/%{k8s_commit}/%{k8s_repo}-%{k8s_shortcommit}.tar.gz
+Source0:        %{name}.tar.gz
+Source1:        https://%{provider_prefix}/archive/%{commit}/%{repo}-%{shortcommit}.tar.gz
```
Comment (mopsfelder) [inline kubernetes/centOS/7/kubernetes.spec]: Why removing `k8s_` prefix from macros in `Source1`?  This changes the actual URL, see:

```rpm-spec
%global provider        github
%global provider_tld    com
%global project	        openshift
%global repo            origin
# https://github.com/openshift/origin
%global provider_prefix	%{provider}.%{provider_tld}/%{project}/%{repo}
%global import_path     k8s.io/kubernetes
%global commit		ef1caba064de975387860175c3138aad432cf356
%global shortcommit	%(c=%{commit}; echo ${c:0:7})

%global openshift_ip    github.com/openshift/origin

%global k8s_provider        github
%global k8s_provider_tld    com
%global k8s_project         kubernetes
%global k8s_repo kubernetes
```

File: kubernetes/centOS/7/kubernetes.spec [hunk 8]
```diff
@@ -575,15 +575,19 @@ BuildRequires: golang >= 1.2-7
 Kubernetes client tools like kubectl
 
 %prep
-%setup -q -n %{k8s_repo}-%{k8s_commit} -T -b 1
+# Use this line if getting tar.gz from git repository. Currently, this source code
+# is downloaded from a tar.gz file in a HTTP server
+#%setup -q -n %{k8s_repo}-%{k8s_commit} -T -b 1
+# -n param value is not %{name} because the tar.gz is not built by us
+%setup -q -n %{repo}-%{commit}
```
Comment (mopsfelder) [inline kubernetes/centOS/7/kubernetes.spec]: Macro `k8s_repo` is different from `repo`, same to `k8s_commit` and `commit`.  Can you please explain why changing this?

File: kubernetes/centOS/7/kubernetes.spec [hunk 9]
```diff
@@ -575,15 +575,19 @@ BuildRequires: golang >= 1.2-7
...
 %if 0%{?with_debug}
 %patch3 -p1
 %endif
 # Hack test-cmd.sh to be run with os binaries
 %patch9 -p1
 
-%setup -q -n %{con_repo}-%{con_commit} -T -b 2
 %setup -q -n %{repo}-%{commit}
+%setup -q -n %{con_repo}-%{con_commit} -T -b 2
```
Comment (mopsfelder) [inline kubernetes/centOS/7/kubernetes.spec]: Is there any benefit of inverting the order of these two `%setup` macros?

File: linux-firmware/centOS/7/linux-firmware.spec [hunk 10]
```diff
@@ -237,7 +237,7 @@ contained inside the provided LICENSE file. Please read it carefully.
 
 %prep
 # Setup Source0 and additional Source1
-%setup -q -c -n linux-firmware-%{checkout}
+%setup -q -n %{name}
```
Comment (mopsfelder) [inline linux-firmware/centOS/7/linux-firmware.spec]: Why `-c` was removed?

File: libseccomp/centOS/7/libseccomp.spec (hunk 6 above)
Comment (mopsfelder) [inline libseccomp/centOS/7/libseccomp.spec]: I'm sure a Fedora recommendation on that has a lot of background involved.  That's not the point here.

If we proceed by removing the URL from `Source`, shouldn't it be `Source: %{name}.tar.gz` instead?

File: kubernetes/centOS/7/kubernetes.spec (hunk 9 above)
Comment (alanoe) [inline kubernetes/centOS/7/kubernetes.spec]: kubernets build is broken in this patch because of the archive removal. I'm fixing it. Build code requires a versions git repo branch to test and I was using the same branch of the PR, which automatically updates the PR content. Pls ignore this version of the patch. I'll use a separate temp branch for testing from now on

Comment (mopsfelder) [inline kubernetes/centOS/7/kubernetes.spec]: Thanks for the explanation.  So just make sure this is removed from the pull request to be reviewed.

File: linux-firmware/centOS/7/linux-firmware.spec (hunk 10 above)
Comment (alanoe) [inline linux-firmware/centOS/7/linux-firmware.spec]: -c was required to create a top-level directory for the sources when linux-firmware-%{checkout} was used as name, given that Source0 was %{name}.tar.gz. This is no longer needed

File: libseccomp/centOS/7/libseccomp.spec (hunk 6 above)
Comment (alanoe) [inline libseccomp/centOS/7/libseccomp.spec]: MO> I'm sure a Fedora recommendation on that has a lot of background involved. That's not the point here.

You asked me why I changed it and I explained it. I thought that was the point. If we are obliged to follow all Fedora recommendations, just tell me and I'll undo my change.

MO> If we proceed by removing the URL from Source, shouldn't it be Source: %{name}.tar.gz instead?

At the point of this patch, nothing functionally was changed. The archive attribute in libseccomp yaml is set to libseccomp-2.3.1 and the source field in rpm spec should point to the same name. In a later patch in this same PR, yaml and rpm spec are simultaneously updated.

Comment (olavphibm) [inline libseccomp/centOS/7/libseccomp.spec]: All of our spec files have the URL removed from the main source, leaving only the file name. We've done it this way because the tar.gz files are created from the git repo, and not downloaded directly. I'm not completely against using an URL, but that would have to be done for all packages and we'd have to check if the URLs indeed lead to the file used.

**Comment by olavphibm** [comment]
start build

File: noVNC/noVNC.yaml [hunk 11]
```diff
@@ -5,7 +5,6 @@ Package:
       src: 'https://github.com/kanaka/noVNC.git'
       branch: 'master' #v0.5.1
       commit_id: 'fc00821eba469641c6c94706726c3d78e46460a2'
-      archive: 'novnc'
```
Comment (olavphibm) [inline noVNC/noVNC.yaml]: By removing this, the archive name has changed to `noVNC`. I suggest keeping the archive attribute to avoid further modifications in the spec file.

Comment (alanoe) [inline noVNC/noVNC.yaml]: As I mentioned in the past, the 'archive' attrib is never needed for the main source file. I have renamed the 'noVNC' diretory and yaml/spec files to 'novnc' and the problem is fixed.

Comment (olavphibm) [inline noVNC/noVNC.yaml]: Building is still broken for this specific commit. I prefer to try to keep all commits buildable, and for that you'd have to rebase the noVNC rename before this commit.

File: gcc/centOS/7/gcc.spec [hunk 12]
```diff
@@ -94,9 +94,9 @@ License: GPLv3+ and GPLv3+ with exceptions and GPLv2+ with exceptions and LGPLv2
 Group: Development/Languages
 # The source for this package was pulled from upstream's vcs.  Use the
 # following commands to generate the tarball:
-# svn export svn://gcc.gnu.org/svn/gcc/branches/redhat/gcc-4_8-branch@%{SVNREV} gcc-%{version}-%{DATE}
-# tar cf - gcc-%{version}-%{DATE} | bzip2 -9 > gcc-%{version}-%{DATE}.tar.bz2
-Source0: gcc-%{version}-%{DATE}.tar.bz2
+# svn export svn://gcc.gnu.org/svn/gcc/branches/redhat/gcc-4_8-branch@%{SVNREV} gcc
+# tar zcf gcc.tar.gz gcc
+Source0: gcc.tar.gz
```
Comment (olavphibm) [inline gcc/centOS/7/gcc.spec]: This should be %{name} to be consistent with other specs.

Comment (alanoe) [inline gcc/centOS/7/gcc.spec]: Done
